### PR TITLE
Refactor toaster/form submit feedback | Closes #2461, #2444

### LINKED ISSUE #2461: Integrate Apollo Update State with RFF / Espresso Form Submission and Toaster to Provide User Feedback
Currently when a user is editing an entity using the full edit entity form, such as `domain/eventEditor/ui/datetimes/dateForm` there is no feedback when the user clicks submit. The form seems to just sit there doing nothing for a few seconds before closing, after which nothing happens for another few seconds or so before finally a success (or error) notification appears in the bottom right of the screen via the toaster.

The AntD Modal component that the edit forms load into allows for a "loading" state spinner to be displayed in the submit button after it is clicked. Ideally this should be activated as soon as submit is clicked so that the user is aware that things are actually happening under the hood.

As well, Apollo should be triggering some kind of "update in progress" notification via the toaster, but that may not be happening due to #2444

Bottom line is that we need to provide some better feedback to users when form submissions occur.

### PR DESCRIPTION
**Pull request by manzoorwanijk**
This PR:
- Refactors toasters to simplify the API.
- Updates the toaster consumers to use the new API.
- Closes #2461, #2444

With the new API, here is how to use the toaster:

```ts
const toaster = useToaster();

// generate a key for loading toaster (opional)
const loadingToastKey = toaster.generateKey();
toaster.loading({
	key: loadingToastKey,
	message: 'Loading, please wait...',
});

// use the above key to dismiss the toaster
toaster.dismiss(loadingToastKey);

// likewise other toasters with custom options, same as AntD notification options
toaster.success({ message: 'Loaded successfully' });
```

### CODE REVIEW

File: assets/src/application/services/toaster/useToastManager.tsx [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+	const closeToast = (key: string): void => {
+		notification.close(key);
+		setToastStack((currentStack) => currentStack.filter((toastKey) => toastKey !== key));
+	};
+
+	const closeIfAtLimit = (): void => {
+		if (toastStack.length === maxToasts) {
+			// remove the oldest toast.
+			// closeToast(toastStack[toastStack.length - 1]);
```
Comment (manzoorwanijk) [inline assets/src/application/services/toaster/useToastManager.tsx]: I have temporarily disabled it, just to see how it behaves without limit.

File: assets/src/domain/eventEditor/services/apollo/queries/tickets/useFetchTickets.ts [hunk 2]
```diff
@@ -17,37 +16,35 @@ const useFetchTickets = (skipFetch: boolean = null): FetchEntitiesResult<Tickets
...
-	useEffect(() => {
-		if (!initialized && !loading) {
-			setInitialized(true);
+		if (loading && !toaster.isToastOpen(loadingToastKey.current)) {
+			toaster.loading({
+				key: loadingToastKey.current,
+				message: 'initializing tickets',
+			});
+		} else if (!loading) {
+			toaster.dismiss(loadingToastKey.current);
```
Comment (nicolad) [inline assets/src/domain/eventEditor/services/apollo/queries/tickets/useFetchTickets.ts]: Can this be somehow extracted, I've seen similar code being repeated above.

Comment (manzoorwanijk) [reply assets/src/domain/eventEditor/services/apollo/queries/tickets/useFetchTickets.ts]: Yes, sure

File: assets/src/application/services/toaster/useToastManager.tsx [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	return {
+		success: getToast('success'),
+		info: getToast('info'),
+		warning: getToast('warning'),
+		error: getToast('error'),
+		loading: getToast('loading'),
+		generateKey: uuidv4,
+		dismiss: closeToast,
+		isToastOpen,
+	};
```
Comment (tn3rb) [inline assets/src/application/services/toaster/useToastManager.tsx]: wow... very nice

File: assets/src/application/services/apollo/mutations/useMutationHandler.ts [hunk 4]
```diff
@@ -54,19 +54,27 @@ const useMutationHandler: MutationHandler = (getMutationOptions) => {
...
 				// e.g. "updating datetime"
 				return `${verb} ${typeName.toLowerCase()}`;
 			};
 
+			/**
+			 * @param {string} mutationType Type of mutation - CREATE|UPDATE|DELETE
+			 */
+			const getToasterKey = (mutationType: MutationType): string => {
+				return `${mutationType}:${id}`;
+			};
```
Comment (tn3rb) [inline assets/src/application/services/apollo/mutations/useMutationHandler.ts]: very cool!

File: assets/src/domain/eventEditor/services/apollo/queries/tickets/useFetchTickets.ts (hunk 2 above)
Comment (tn3rb) [reply assets/src/domain/eventEditor/services/apollo/queries/tickets/useFetchTickets.ts]: I concur!

Comment (manzoorwanijk) [reply assets/src/domain/eventEditor/services/apollo/queries/tickets/useFetchTickets.ts]: Done

Comment (tn3rb) [reply assets/src/domain/eventEditor/services/apollo/queries/tickets/useFetchTickets.ts]: and pushed?

File: assets/src/domain/eventEditor/services/apollo/queries/datetimes/useFetchDatetimes.ts [hunk 5]
```diff
@@ -29,16 +29,10 @@ const useFetchDatetimes = (): FetchEntitiesResult<DatetimesList> => {
 		},
 	});
 
+	useLoadingToast({ loading, message: 'initializing datetimes', toastKey: loadingToastKey.current });
```
Comment (tn3rb) [inline assets/src/domain/eventEditor/services/apollo/queries/datetimes/useFetchDatetimes.ts]: ya that works!

**Pull request merged**